File: dexfile/src/main/java/com/github/netomi/bat/dexfile/FieldID.java
/*
 *  Copyright (c) 2020 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.github.netomi.bat.dexfile;

import com.github.netomi.bat.dexfile.io.DexDataInput;
import com.github.netomi.bat.dexfile.io.DexDataOutput;
import com.github.netomi.bat.util.Preconditions;

import java.util.Objects;

import static com.github.netomi.bat.dexfile.DexConstants.NO_INDEX;

/**
 * A class representing a field id item inside a dex file.
 *
 * @see <a href="https://source.android.com/devices/tech/dalvik/dex-format#field-id-item">field id item @ dex format</a>
 *
 * @author Thomas Neidhart
 */
@DataItemAnn(
    type          = DexConstants.TYPE_FIELD_ID_ITEM,
    dataAlignment = 4,
    dataSection   = false
)
public class FieldID
extends      DataItem
{
    private int classIndex; // ushort
    private int typeIndex;  // ushort
    private int nameIndex;  // uint;

    public static FieldID of(int classIndex, int nameIndex, int typeIndex) {
        Preconditions.checkArgument(classIndex >= 0, "class index must be non negative");
        Preconditions.checkArgument(nameIndex  >= 0, "name index must be non negative");
        Preconditions.checkArgument(typeIndex  >= 0, "type index must be non negative");
        return new FieldID(classIndex, nameIndex, typeIndex);
    }

    public static FieldID readContent(DexDataInput input) {
        FieldID fieldID = new FieldID();
        fieldID.read(input);
        return fieldID;
    }

    private FieldID() {
        this(NO_INDEX, NO_INDEX, NO_INDEX);
    }

    private FieldID(int classIndex, int nameIndex, int typeIndex) {
        this.classIndex = classIndex;
        this.nameIndex  = nameIndex;
        this.typeIndex  = typeIndex;
    }

    public int getClassIndex() {
        return classIndex;
    }

    public TypeID getClassTypeID(DexFile dexFile) {
        return dexFile.getTypeID(classIndex);
    }

    public String getClassType(DexFile dexFile) {
        return getClassTypeID(dexFile).getType(dexFile);
    }

    public int getTypeIndex() {
        return typeIndex;
    }

    public String getType(DexFile dexFile) {
        return dexFile.getTypeID(typeIndex).getType(dexFile);
    }

    public int getNameIndex() {
        return nameIndex;
    }

    public String getName(DexFile dexFile) {
        return dexFile.getStringID(nameIndex).getStringValue();
    }

    @Override
    protected void read(DexDataInput input) {
        input.skipAlignmentPadding(getDataAlignment());
        classIndex = input.readUnsignedShort();
        typeIndex  = input.readUnsignedShort();
        nameIndex  = input.readInt();
    }

    @Override
    protected void write(DexDataOutput output) {
        output.writeAlignmentPadding(getDataAlignment());
        output.writeUnsignedShort(classIndex);
        output.writeUnsignedShort(typeIndex);
        output.writeInt(nameIndex);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        FieldID other = (FieldID) o;
        return classIndex == other.classIndex &&
               typeIndex  == other.typeIndex &&
               nameIndex  == other.nameIndex;
    }

    @Override
    public int hashCode() {
        return Objects.hash(classIndex, typeIndex, nameIndex);
    }

    @Override
    public String toString() {
        return String.format("FieldID[classIdx=%d,typeIdx=%d,nameIdx=%d]", classIndex, typeIndex, nameIndex);
    }
}

File: dexfile/src/main/kotlin/com/github/netomi/bat/dexfile/FieldID.kt
/*
 *  Copyright (c) 2020 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.github.netomi.bat.dexfile

import com.github.netomi.bat.dexfile.io.DexDataInput
import com.github.netomi.bat.dexfile.io.DexDataOutput
import com.github.netomi.bat.dexfile.util.DexType
import com.github.netomi.bat.dexfile.visitor.PropertyAccessor
import com.github.netomi.bat.dexfile.visitor.ReferencedIDVisitor
import java.util.*

/**
 * A class representing a field id item inside a dex file.
 *
 * @see <a href="https://source.android.com/devices/tech/dalvik/dex-format#field-id-item">field id item @ dex format</a>
 */
@DataItemAnn(
    type          = TYPE_FIELD_ID_ITEM,
    dataAlignment = 4,
    dataSection   = false)
class FieldID private constructor(classIndex: Int = NO_INDEX,
                                  nameIndex:  Int = NO_INDEX,
                                  typeIndex:  Int = NO_INDEX): DataItem() {

    var classIndex: Int = classIndex
        private set

    var nameIndex: Int = nameIndex
        private set

    var typeIndex: Int = typeIndex
        private set

    fun getClassTypeID(dexFile: DexFile): TypeID {
        return dexFile.getTypeID(classIndex)
    }

    fun getClassType(dexFile: DexFile): DexType {
        return getClassTypeID(dexFile).getType(dexFile)
    }

    fun getTypeID(dexFile: DexFile): TypeID {
        return dexFile.getTypeID(typeIndex)
    }

    fun getType(dexFile: DexFile): DexType {
        return getTypeID(dexFile).getType(dexFile)
    }

    fun getName(dexFile: DexFile): String {
        return dexFile.getStringID(nameIndex).stringValue
    }

    fun getFullExternalFieldSignature(dexFile: DexFile): String {
        return buildString {
            append(getClassType(dexFile).toExternalClassName())
            append('.')
            append(getName(dexFile))
            append(':')
            append(getType(dexFile).toExternalType())
        }
    }

    /**
     * This item is considered to be empty only if it's an uninitialized instance.
     */
    override val isEmpty: Boolean
        get() = classIndex == NO_INDEX

    override fun read(input: DexDataInput) {
        classIndex = input.readUnsignedShort()
        typeIndex  = input.readUnsignedShort()
        nameIndex  = input.readInt()
    }

    override fun write(output: DexDataOutput) {
        output.writeUnsignedShort(classIndex)
        output.writeUnsignedShort(typeIndex)
        output.writeInt(nameIndex)
    }

    internal fun referencedIDsAccept(dexFile: DexFile, visitor: ReferencedIDVisitor) {
        visitor.visitTypeID(dexFile, PropertyAccessor(::classIndex))
        visitor.visitStringID(dexFile, PropertyAccessor(::nameIndex))
        visitor.visitTypeID(dexFile, PropertyAccessor(::typeIndex))
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as FieldID

        return classIndex == other.classIndex &&
               nameIndex  == other.nameIndex &&
               typeIndex  == other.typeIndex
    }

    override fun hashCode(): Int {
        return Objects.hash(classIndex, nameIndex, typeIndex)
    }

    override fun toString(): String {
        return "FieldID[classIdx=${classIndex},nameIdx=${nameIndex},typeIdx=${typeIndex}]"
    }

    companion object {
        fun of(classIndex: Int, nameIndex: Int, typeIndex: Int): FieldID {
            require(classIndex >= 0) { "class index must not be negative" }
            require(nameIndex >= 0) { "name index must not be negative" }
            require(typeIndex >= 0) { "type index must not be negative" }
            return FieldID(classIndex, nameIndex, typeIndex)
        }

        internal fun read(input: DexDataInput): FieldID {
            val fieldID = FieldID()
            fieldID.read(input)
            return fieldID
        }
    }
}